snakemake/executors.py

```diff
@@ -2052,3 +2059,605 @@ def run_wrapper(
+            logger.warning(
+                "Singularity requires additional capabilities that "
+                "aren't yet supported for standard Docker runs."
+            )
+
+        # Note that locations might need to be filtered down
+        parent = "projects/%s/locations/*" % self.project
+        operation = pipelines.run(parent=parent, body=body)
+
+        # 403 will result if no permission to use pipelines or project
+        result = self._retry_request(operation)
+
+        # The jobid is the last number of the full name
+        jobid = result["name"].split("/")[-1]
+
+        # Give some logging for how to get status
+        logger.info(
+            "Get status with:\n"
+            "gcloud config set project {project}\n"
+            "gcloud beta lifesciences operations describe {jobid}\n"
+            "gcloud beta lifesciences operations list".format(
+                project=self.project, jobid=jobid
+            )
+        )
+
+        self.active_jobs.append(
+            GoogleLifeScienceJob(job, result["name"], jobid, callback, error_callback)
+        )
+
+    def _job_was_successful(self, status):
+        """based on a status response (a [pipeline].projects.locations.operations.get
+           debug print the list of events, return True if all return codes 0
+           and False otherwise (indication of failure). In that a nonzero exit
+           status is found, we also debug print it for the user.
+        """
+        success = True
+
+        # https://cloud.google.com/life-sciences/docs/reference/rest/v2beta/Event
+        for event in status["metadata"]["events"]:
+
+            logger.info(event["description"])
+
+            # Does it always result in fail for other failure reasons?
+            if "failed" in event:
+                success = False
+                action = event.get("failed")
+                logger.error("{}: {}".format(action["code"], action["cause"]))
+
+            elif "unexpectedExitStatus" in event:
+                action = event.get("unexpectedExitStatus")
+
+                if action["exitStatus"] != 0:
+                    success = False
+
+                    # Provide reason for the failure (desc includes exit code)
+                    msg = "%s" % event["description"]
+                    if "stderr" in action:
+                        msg += ": %s" % action["stderr"]
+                        logger.error(msg)
+
+        return success
+
+    def _retry_request(self, request, timeout=2, attempts=3):
+        """The Google Python API client frequently has BrokenPipe errors. This
+           function takes a request, and executes it up to number of retry,
+           each time with a 2* increase in timeout.
+
+           Parameters
+           ==========
+           request: the Google Cloud request that needs to be executed
+           timeout: time to sleep (in seconds) before trying again
+           attempts: remaining attempts, throw error when hit 0
+        """
```

Comment:
@moschetti is there a chance to fix this upstream?

Reply:
What do you mean upstream? @moschetti can comment, but I've worked with a lot of Google APIs, and the retry functionality (at least afaik) doesn't come with that, there is always a suggestion to implement some kind of [exponential backoff](https://cloud.google.com/iot/docs/how-tos/exponential-backoff). And when I haven't, it leads to a lot of random errors because there is the occasional error, even with Google.

Reply:
as mentioned, the retry and backoff gets implemented here at the client level. that's above my pay grade, but is the general behavior of all the exposed APIs so that users can handle error behavior in a way that's appropriate for their use case.

Reply:
@johanneskoester I think what I've implemented is a reasonable start (and perhaps we might want to adjust the timeout original value (2) or the retry count (3)? What we basically do is submit the operation of the bat (Google discovery requests have an .execute() function that allows us to pass the request and then execute). And then in the case of a BrokenPipe (the error I was seeing a few times) we run the function again, of course multiplying the timeout so we wait a little longer, and subtracting one from the retry count:

```python
        except BrokenPipeError as ex:
            if attempts > 0:
                sleep(timeout)
                return self._retry_request(request, timeout * 2, attempts - 1)
            raise ex
```
Also notice in the above that's where we wait (the sleep) and then in the case that we've exhausted the retry count, we raise the exception. I think this should work (I haven't hit issues with it yet) and maximally we might need to look for other exception types (haven't seen any yet).

Reply:
Sounds good to me. I just wanted to check if Google wants us to do it like that.